### DISCUSSION TITLE
Annotato pipeline issues

### DISCUSSION THREAD

**Discussion by Antonio-Longo** [ERGA-consortium/pipelines#49] [Q&A] [answered]
Hi again everybody,

I am trying to run the Annotato pipeline. As suggested in the prerequisites, I installed the following:

nextflow v24.04.4.5917
singularity v3.5.3 (go v1.23.0)
conda and mamba (I used the latter to create a dedicated environment)
docker v27.1.1

To test it, I selected a draft MAG genome.fa file (4.5 GB), without any other input data. I so used the following command:

`nextflow run main.nf --genome /home/path/to/.fa --species "xxx xxx"
`

that gave me the attached error and nextflow.log.

I am reading about the suggestion about singularity.pullTimeout (for instance changing it as described here: https://www.nextflow.io/docs/latest/config.html) 

Do you have any suggestion,

Thank you and best regards!

Antonio

![image](https://github.com/user-attachments/assets/e2bf1f71-61b7-4a37-93eb-0fbb68189ce1)

[error.log](https://github.com/user-attachments/files/16846951/error.log)

**Comment by gkanogiannis** [comment]
Try with "--singularity.pullTimeout 60m" at the end of your nextflow command

**Comment by Antonio-Longo** [comment]
still the same error, here the new log:

[nextflow.log](https://github.com/user-attachments/files/16847221/nextflow.log)


thanks

**Comment by gkanogiannis** [comment]
Ops sorry, try to use single dash, like -singularity.pullTimeout 60m

**Reply by gkanogiannis** [reply]
Actually, this would not work either.

Try to pass the config value through a custom config file.
Create a file `my_config` with contents 
`
singularity {
        pullTimeout = '60m'
}
`

and append `-c my_config` at the end of your command. I hope this works.

**Reply by Antonio-Longo** [reply]
still the same error, i tried also higher values of pullTimeout.

![image](https://github.com/user-attachments/assets/1bc7f433-cff3-4490-bf61-b0e94587e27e)

**Reply by gkanogiannis** [reply]
What is the version of singularity in your system?

**Reply by Antonio-Longo** [reply]
singularity v3.5.3 (go v1.23.0)

**Reply by gkanogiannis** [reply]
That's a VERY old version, more than 4 years old, that does not support the v2 SIF images. This support came on with singularity v3.6
Upgrade your singularity to a more recent version (preferably a recent v4.x)

**Reply by Antonio-Longo** [reply]
ah I missed this. 
So it works with v3.9 (tell me if it is still the case to upgrade to the v4), but it give me another problem at:

 ANNOTATO:GENOME_MASKING:FILTER_REPEAT:MAKEBLASTDB (1)

![image](https://github.com/user-attachments/assets/ecf44241-805d-4d9f-ba05-6b1d32a996b7)
[nextflow.log](https://github.com/user-attachments/files/16850269/nextflow.log)


Thank you very much for your help!

**Reply by gkanogiannis** [reply]
Annotato pipeline by default runs the slurm engine (that's why your error message fails in finding sbatch).

Apparently you run the pipeline in a common PC.

Use `-profile local,singularity` at the end of your nextflow command.

**Reply by Antonio-Longo** [reply]
yes i am working on local (the readme file tells about problems on HPC). 

with the local flag it is going further, but still crashes  at STRINGTIE step

![image](https://github.com/user-attachments/assets/0a98fba9-4015-4278-813b-13928dc594fd)
[nextflow.log](https://github.com/user-attachments/files/16850438/nextflow.log)

your help is precious

**Reply by gkanogiannis** [reply]
Ok, it seems there is a bug in annotato when you only provide genome as input.
I submitted a PR with the fix, but until it is merged you can apply the fix manually.
Inside the annotato folder, edit the file `workflows/annotato.nf` in line 282 to look like 
```nextflow
if(ch_short_rna_bam_stringtie != "EMPTY" || ch_long_rna_bam_stringtie != "EMPTY") {
    STRINGTIE ( ch_short_rna_bam_stringtie, ch_long_rna_bam_stringtie )
    ch_versions = ch_versions.mix(STRINGTIE.out.versions.ifEmpty(null))
    ch_stringtie_gtf = STRINGTIE.out.gtf
}
```

**Reply by Antonio-Longo** [reply]
step by step it is working. Now the error is:

![image](https://github.com/user-attachments/assets/28c45129-dc64-4efe-b24f-67f5123108ff)

is there a way to either adapt the chunck of code of GENOME_MASKING or the main command to set the number of threads or cores it uses?

**Reply by gkanogiannis** [reply]
Append `--max_cpus 16 --max_cores 16` in your nextflow command

**Reply by Antonio-Longo** [reply]
It seems that we are half way, thanks for this.

This is happening at protexcluder step:

![image](https://github.com/user-attachments/assets/e2d71744-ce93-44e9-b0ac-7552297a2462)
[nextflow.log](https://github.com/user-attachments/files/16862078/nextflow.log)

**Reply by gkanogiannis** [reply]
It seems protexcluder cannot access some output files.
I am taking wild guesses now, is there enough space in the filesystem you are running this?
Try adding `--publish_dir_mode symlink` at the end of the command.

**Reply by Antonio-Longo** [reply]
I am now working on 500 GB free space with a "reduced" version of the genome (1mb, ca. 300 contigs). 

[nextflow.log](https://github.com/user-attachments/files/16868515/nextflow.log)
![image](https://github.com/user-attachments/assets/3078ee57-0f88-4eb1-b70d-267cbc45591b)

**Reply by gkanogiannis** [reply]
Yeah, I get the same error when running with a small genome.
There is another bug, but until the fix is merged you can apply it manually.
Edit file `modules/local/repeatmasker/RepeatMasker.nf` and add after line 27 
```
[ -f genome.fasta.cat.gz ] || gzip genome.fasta.cat
```

**Reply by Antonio-Longo** [reply]
it still give me the problem:


[nextflow.log](https://github.com/user-attachments/files/16870150/nextflow.log)

I am trying with other genomes (smaller than 10 MB), but still it fails


thanks

**Reply by Antonio-Longo** [reply]
Dear all,

I am still trying to let the annotato pipeline work, at the moment I am trying to annotate a single euk MAG (12 MB) with the following command:

`nextflow run main.nf --genome /path/to/mag.fa --species 'unk euk' -profile local,singularity --max_cpus 14 --max_cores 14`

![image](https://github.com/user-attachments/assets/69cc3b61-7f67-49da-8662-aedf3db83c09)
[nextflow.log](https://github.com/user-attachments/files/16909346/nextflow.log)

thanks for your time, and help.

**Reply by gkanogiannis** [reply]
Try to rerun the pipeline, append `-resume` in the command line.

**Reply by Antonio-Longo** [reply]
I did managed to install 4.x singularity version.

Rerunning with -resume worked and i managed to go further in the pipeline steps, still it gives me this error:

![image](https://github.com/user-attachments/assets/0d67dd9f-19bd-4aa3-bb05-4b409cf9a4d1)

It seems that there is a proeblem mounting directories in the container environment used by funnanotate... but I could be wrong.

[log.txt](https://github.com/user-attachments/files/17130059/log.txt)

**Reply by gkanogiannis** [reply]
Pull the latest version. We removed those mounts that were creating issues.

Edit: it has not been merged yet. Try to manually edit `annotation/nextflow/modules/local/funannotate/FunannotatePredict.nf` and change line `containerOptions = "--bind /env:/env --bind ${params.tmpdir}:/opt/databases --bind /mnt:/mnt"` to `containerOptions = "--bind ${params.tmpdir}:/opt/databases"`

**Reply by Antonio-Longo** [reply]
almost there!!

![image](https://github.com/user-attachments/assets/0c12fe98-bd33-481d-b87e-f2db9663aabb)

is there the possibility of reducing the required gene models to run BUSCO, or I should try with "bigger" MAGs (now it is around 10 MB size)... I found something about --min_training_models 

[log.txt](https://github.com/user-attachments/files/17131233/log.txt)

**Reply by gkanogiannis** [reply]
We added a parameter in the pipeline for controlling this value. It is not yet pushed.
@tbrown91 and @diegomics are reviewing the relevant PR.
Until then, you can manually edit `annotation/nextflow/modules/local/funannotate/FunannotatePredict.nf` and append `--min_training_models 150` in the funannotate predict command.

**Reply by Antonio-Longo** [reply, marked as answer]
So it is finally working! I don't know how to thank you for your time and work. Now i will try it with RNAseq...

![image](https://github.com/user-attachments/assets/06db14e7-8742-43ab-9b16-e00ddcce2cde)

Thanks, thanks and thanks again @gkanogiannis

**Comment by tbrown91** [comment]
Thanks @gkanogiannis for your fix https://github.com/ERGA-consortium/pipelines/pull/50

@Antonio-Longo can you pull this version and give it a try?

**Reply by Antonio-Longo** [reply]
it is working, still some debug to run the pipeline!

**Reply by Antonio-Longo** [reply]
Hi, i just pulled the pipeline and tried it. Unfortunately it crashed with the following error:

ERROR ~ Failed to publish file: /home/alongo/Desktop/aperti/NEREA/annotato/work/b7/3c1640a510e3a46670eecb5eb3b97f/RM_41.SatOct261202092024; to: /home/alongo/Desktop/aperti/NEREA/annotato/02.Repeatmasker/RepeatModeler/RM_41.SatOct261202092024 [copy] -- See log file for details

 -- Check '.nextflow.log' file for details

and from the log it seems that in funannotate to fail:

[nextflow.log.txt](https://github.com/user-attachments/files/17530609/nextflow.log.txt)

I can use -resume as @gkanogiannis suggested, but it would be great to have this working...



Moreover, if i resume the work the in the final .gff3 output does of funannotate is missing the proper annotation (the file reports hypothetical protein for each entry, and if i search one of them on uniprot there is a match). Maybe using a different busco db (but none of the scaffold from the analysis was annotated with default eukaryotes).

Thanks again for your help

**Reply by gkanogiannis** [reply]
Try maybe `--publish_dir_mode symlink`

**Reply by Antonio-Longo** [reply]
ok this is helpful to close the pipeline, but i just noticed that the annotation part of funannotate is missing. Do you think it will be created or am I missing something?

however with interproscan tool i got pretty good results!

thanks again
